src/Microsoft.AspNetCore.Routing/Matchers/AsciiKeyedJumpTable.cs

```diff
@@ -0,0 +1,161 @@
+    {
+        public static bool TryCreate(
+            int @default,
+            int exit,
+            List<(string text, int destination)> entries,
+            out JumpTable result)
+        {
+            result = null;
+
+            // First we group string by their uppercase letter. If we see a string
+            // that starts with a non-ASCII letter 
+            var map = new Dictionary<char, List<(string text, int destination)>>();
+
+            for (var i = 0; i < entries.Count; i++)
+            {
+                if (entries[i].text.Length == 0)
+                {
+                    return false;
+                }
+
+                if (!IsAscii(entries[i].text))
+                {
+                    return false;
+                }
+                
+                var first = ToUpperAscii(entries[i].text[0]);
+                if (first < 'A' || first > 'Z')
+                {
+                    // Not a letter
+                    return false;
+                }
+
+                if (!map.TryGetValue(first, out var matches))
+                {
+                    matches = new List<(string text, int destination)>();
+                    map.Add(first, matches);
+                }
+
+                matches.Add(entries[i]);
+            }
+
+            var next = 0;
+            var ordered = new (string text, int destination)[entries.Count];
+            var indexes = new int[26 * 2];
+            for (var i = 0; i < 26; i++)
+            {
+                indexes[i * 2] = next;
+
+                var length = 0;
+                if (map.TryGetValue((char)('A' + i), out var matches))
+                {
+                    length += matches.Count;
+                    for (var j = 0; j < matches.Count; j++)
+                    {
+                        ordered[next++] = matches[j];
+                    }
+                }
+
+                indexes[i * 2 + 1] = length;
+            }
+
+            result = new AsciiKeyedJumpTable(@default, exit, ordered, indexes);
+            return true;
+        }
+
+        private readonly int _default;
+        private readonly int _exit;
+        private readonly (string text, int destination)[] _entries;
+        private readonly int[] _indexes;
+
+        private AsciiKeyedJumpTable(
+            int @default,
+            int exit,
+            (string text, int destination)[] entries,
+            int[] indexes)
+        {
+            _default = @default;
+            _exit = exit;
+            _entries = entries;
+            _indexes = indexes;
+        }
+
+        public unsafe override int GetDestination(string path, PathSegment segment)
+        {
+            if (segment.Length == 0)
+            {
+                return _exit;
+            }
+
+            var c = path[segment.Start];
+            if ((c & ~0x7F) != 0)
+            {
+                return _default;
+            }
+
+            c = (char)(c & 0x5F);
```

Comment:
Move value to const variable and/or comment explaining what it is

Reply:
Use `ToUpperAscii`?

Reply:
Sure